backend/compact-connect/lambdas/python/provider-data-v1/handlers/providers.py

```diff
@@ -153,3 +171,111 @@ def get_provider_ssn(event: dict, context: LambdaContext):  # noqa: ARG001 unuse
         return {
             'ssn': ssn,
         }
+
+
+def _ssn_rate_limit_exceeded(user_id: str, provider_id: str, compact: str) -> bool:
+    """Check if the user has exceeded the SSN rate limit.
+
+    :param user_id: The Cognito user ID of the staff user
+    :param provider_id: The provider ID being accessed
+    :param compact: The compact being accessed
+    :return: True if rate limit is exceeded, False otherwise
+    """
+    now = config.current_standard_datetime
+    window_start = now - timedelta(hours=24)
+    window_start_timestamp = window_start.timestamp()
+    now_timestamp = now.timestamp()
+
+    # Generate a unique ID for this request
+    # This ensures every request is recorded, even for requests within the same second
+    request_sk = f'TIME#{now_timestamp}#UUID#{uuid.uuid4()}'
+
+    try:
+        # First, record this request in the rate limiting table
+        config.rate_limiting_table.put_item(
+            Item={
+                'pk': 'READ_SSN_REQUESTS',
+                'sk': request_sk,
+                'compact': compact,
+                'provider_id': provider_id,
+                'staffUserId': user_id,
+                'ttl': int(now_timestamp) + 86400,  # 24 hours in seconds
+            }
+        )
+
+        # Check if the global rate limit has been exceeded (more than 15 requests in 24 hours)
+        all_requests = config.rate_limiting_table.query(
+            KeyConditionExpression='pk = :pk AND sk BETWEEN :start_sk AND :end_sk',
+            ExpressionAttributeValues={
+                ':pk': 'READ_SSN_REQUESTS',
+                ':start_sk': f'TIME#{window_start_timestamp}',
+                # Add 1 second to ensure we include all records at the current timestamp
+                ':end_sk': f'TIME#{now_timestamp + 1}',
+            },
+            ConsistentRead=True,
+        )
+
+        global_request_count = len(all_requests['Items'])
+        logger.info(f'Global SSN request count in last 24 hours: {global_request_count}')
+
+        # If there are more than 15 requests globally in the last 24 hours, throttle the entire endpoint
+        if global_request_count > 15:
+            logger.critical(
+                'Global SSN rate limit exceeded, throttling endpoint',
+                global_request_count=global_request_count,
+                current_request_user_id=user_id,
+                current_request_compact=compact,
+            )
+
+            # Set the lambda's reserved concurrency to 0 to throttle the endpoint
+            try:
+                config.lambda_client.put_function_concurrency(
+                    FunctionName=config.current_lambda_name, ReservedConcurrentExecutions=0
```

Comment:
You can get this from the `context` object as well, FWIW:
https://docs.aws.amazon.com/lambda/latest/dg/python-context.html